music21/abcFormat/testFiles.py

```diff
@@ -612,8 +676,116 @@ def testBasic(self):
             if title == 'D Fragment':
                 sharps = s.parts[0].keySignature.sharps
                 self.assertEqual(sharps, 7, 'C# key signature should be parsed as 7 sharps')
+                
+    def testAbc21(self):
+        from music21 import abcFormat, note
+        from music21.abcFormat import translate
+        from music21.musicxml import m21ToXml
 
+        af = abcFormat.ABCFile(abcVersion="2.1")
+
+        gex = m21ToXml.GeneralObjectExporter()
         
+        files = [carryThrough]
+        for i, tf in enumerate(files):
+            ah = af.readstr(tf)
+            title = ah.getTitle()
+            environLocal.printDebug([title])
+            s = translate.abcToStreamScore(ah)
+            try:
+                unused_out = gex.parse(s)
+            except UnicodeDecodeError as ude:
+                environLocal.warn('About to fail on ABC file #{}'.format(i))
+                raise ude
+            if title == 'Through Measure':
+                notes = s.flat.getElementsByClass(note.Note)
+                cSharp = notes[3]
+                cThrough = notes[5]
+                self.assertEqual(cSharp.pitch.midi, cThrough.pitch.midi, "Sharp does not carry through measure")
+                bFlat = notes[4]
+                bLast = notes[7]
+                self.assertEqual(bFlat.pitch.midi, bLast.pitch.midi, "Flat does not carry through measure")
+                bNat = notes[10]
+                bNatNext = notes[11]
+                self.assertEqual(bNat.pitch.midi, bNatNext.pitch.midi, "Natural does not carry through measure")
+                self.assertEqual(notes[12].pitch.midi, 73, "Sharp does not carry through measure")
+                self.assertEqual(notes[13].pitch.midi, 72, "Natural is ignored")
+                self.assertEqual(notes[14].pitch.midi, 72, "Natural does not carry through measure")
+                self.assertEqual(notes[16].pitch.midi, 72, "Sharp carries over measure incorrectly")
+                self.assertEqual(notes[17].pitch.midi, 74, "Sharp (D5) carries over measure incorrectly")
+                self.assertEqual(notes[18].pitch.midi, 78, "Natural (F5) carries over measure incorrectly")
+            # elif title == 'Tie Through Measure':
+                # FIXME: Carrying an accidental into the next measure with a tie doesn't work.
+                # notes = s.flat.getElementsByClass(note.Note)
+                # self.assertEqual(notes[4].pitch.midi, 70, "Tied note loses it sharp")
+                # self.assertEqual(notes[6].pitch.midi, 69, "Tied-over sharp persists past the tie")
+
+    def testAbc21Directive(self):
+        from music21 import abcFormat, note
+        from music21.abcFormat import translate
+        from music21.musicxml import m21ToXml
+
+        af = abcFormat.ABCFile()
+
+        gex = m21ToXml.GeneralObjectExporter()
+        
+        files = [directiveCarryPitch, directiveCarryOctave, directiveCarryNot]
+        for i, tf in enumerate(files):
```

Comment:
instead of doing this with a for loop, do a separate test method for each file please.

Reply:
Agreed.